### DISCUSSION TITLE
The new iterator interface is not what I expected

### DISCUSSION THREAD

**Discussion by tpihl** [nats-io/nats.go#1283] [Ideas]
Trying out the new jetstream interface, I got stuck on the iterator pattern.

I was expecting an interface follow bufio.Scanner where Scan returns bool and have a getter for Error and Msg.

What is the benefit of the the current implementation that i fail to see? I think it's harder to read when the error is mixed into the happy-flow. It will stop the loop anyhow so why not follow std lib and allow the simple 'for it.Next(){}'?

/T

**Comment by piotrpio** [comment]
Hey @tpihl, thanks for the feedback.

There is a difference between the core of how `bufio.Scanner{}` and `Messages()` method work - scanner is meant to be finite - thus `Scan()` returning a bool to signal completion - either with error, or successfully (EOF).

Iterating over messages by `Next()` is by design infinite, so returning a bool there makes little sense. What I was thinking about is returning just a `jetstream.Msg` from `Next()` and having `Err()` on `MessagesContext` interface (similarly to `Scanner`). The problem with this approach is, you can either have terminal errors (like e.g. `ErrConsumerDeleted`) or non-terminal error (`ErrNoHeartbeat`), which you possibly may want to observe. So you would have to check for the error on each retrieved message either way.

Ultimately, while it does not follow what `bufio.Scanner{}` does, the happy path scenario is I think quite straightforward:
```go
	for {
		msg, err := it.Next()
		if err != nil {
			// handle error
		}
		msg.Ack()
	}
```

**Comment by tpihl** [comment]
Thank you for replying @piotrpio!

Let me try to show my perspective, since maybe now I understand the difference better.

IMHO the consumer IS finite. It can be terminated by removing the consumer outside the scope of the iterator, it can be terminated by an error and so on. 

I actually think there should be a way to gracefully terminate a consumer, and hence return false without an error (since the termination was intentional).

Seen from the other side;

The bufio scanner is infinite in the same sense, you can use it on stdin without knowing if it ever will finish. But when it does, it will return false and you can check why it stopped. 

/T

**Comment by piotrpio** [comment]
I get your point. From my perspective, removing a consumer from outside the scope of the iterator has to always be an error - what I mean by infinite, there is no scenario in which the iterator would finish and error would be `nil` - we always have to telegraph the reason.

One possible exception to this would be calling `it.Stop()` - this I think could be a use case in which you want to gracefully exit the loop - however, `Stop()` would most likely (from my perspective at least) be usually called in the scope of `Next()` (e.g. in response to some condition on the message, heartbeat error etc.).

One thing which would still be an issue if we do not return `error` from `Next()` - how to telegraph non-terminal errors? E.g. `ErrNoHeartbeat` - this is a general problem with warning-like errors, which most likely you want to ignore (system may come back up), so you may not want to break from the loop immediately - maybe you want to wait until several heartbeats are skipped, or you just want to log it. `bufio.Scanner` does not have this problem - any error terminates the operation. In `Consume()`, we have a custom callback for handling this situation, but here I did not want to mix iterators with async callbacks.

**Reply by tpihl** [reply]
@piotrpio fair points. Personal preference is that if i need to react to a warning, i would not mind register a warning-handler, and if I don't, i don't care. That's only me though, I'm sure.

I actually made another topic (just to not hide answers by switching topic, just about graceful shutdown of consumer. So I consider this reflection closed even though not the way that would suit me (probably only me) the best. I am of course always able to write a wrapper around this api to get things my way ;)

**Comment by tpihl** [comment]
@piotrpio Hi, sorry for re-opening this question, but...

I just saw the pull-request #1312, and I like that iterator way  more than the current consumer iterator. Any chance that the same change can be applied also to that one?

BR/T

**Comment by piotrpio** [comment]
@tpihl I don't like the fact that the way those iterators are handled is different, but there is a big difference - in the methods added in #1312 there is a clear end condition where the channel is closed (and therefore you can safely use `for..range` on the result). With `Messages()`, you don't have such condition, unless there is a terminal error or the iterator is stopped by the user - if it's not the case, it will infinitely wait for new messages.

**Reply by tpihl** [reply]
@piotrpio, Just like a con.Consume() will, right? And you can stop them the same way

**Comment by tpihl** [comment]
however, i realize that this is a dead horse
